driver/pace/driver/report.py

```diff
@@ -0,0 +1,146 @@
+    times: dict
+    SYPD: float
+
+
+def get_experiment_info(
+    experiment_name: str, time_step: int, backend: str, git_hash: str
+) -> Dict[str, Any]:
+    experiment = Experiment(
+        dataset=experiment_name,
+        format_version=3,
+        git_hash=git_hash,
+        timestamp=datetime.now().strftime("%d/%m/%Y %H:%M:%S"),
+        timesteps=time_step,
+        version="python/" + backend,
+    )
+    return experiment
+
+
+def collect_keys_from_data(times_per_step: List[Dict[str, float]]) -> List[str]:
+    """Collects all the keys in the list of dics and returns a sorted version"""
+    keys = set()
+    for data_point in times_per_step:
+        for k, _ in data_point.items():
+            keys.add(k)
+    sorted_keys = list(keys)
+    sorted_keys.sort()
+    return sorted_keys
+
+
+def gather_timing_data(
+    times_per_step: List[Dict[str, float]],
+    comm,
+    root: int = 0,
+) -> Dict[str, Any]:
+    """returns an updated version of  the results dictionary owned
+    by the root node to hold data on the substeps as well as the main loop timers"""
+    is_root = comm.Get_rank() == root
+    keys = collect_keys_from_data(times_per_step)
+    data: List[float] = []
+    timing_info = {}
+    for timer_name in keys:
+        data.clear()
+        for data_point in times_per_step:
+            if timer_name in data_point:
+                data.append(data_point[timer_name])
+
+        sendbuf = np.array(data)
+        recvbuf = None
+        if is_root:
+            recvbuf = np.array([data] * comm.Get_size())
+        comm.Gather(sendbuf, recvbuf, root=0)
+        if is_root:
+            timing_info[timer_name] = TimeReport(
+                hits=0, times=copy.deepcopy(recvbuf.tolist())
+            )
+    return timing_info
+
+
+def write_to_timestamped_json(experiment: Dict[str, Any]) -> None:
+    now = datetime.now()
+    filename = now.strftime("%Y-%m-%d-%H-%M-%S")
+    with open(filename + ".json", "w") as outfile:
+        json.dump(experiment, outfile, sort_keys=True, indent=4)
+
+
+def gather_hit_counts(
+    hits_per_step: List[Dict[str, int]], timing_info: Dict[str, TimeReport]
+) -> Dict[str, Any]:
+    """collects the hit count across all timers called in a program execution"""
+    for data_point in hits_per_step:
+        for name, value in data_point.items():
+            timing_info[name].hits += value
+    return timing_info
+
+
+def get_sypd(timing_info: Dict[str, TimeReport], dt_atmos: float) -> float:
+    if "mainloop" in timing_info:
+        mainloop = np.mean(sum(timing_info["mainloop"].times, []))
+        speedup = dt_atmos / mainloop
+        sypd = 1.0 / 365.0 * speedup
+    else:
+        sypd = -999.0
+    return sypd
+
+
+def make_report(
+    exp_info: Experiment,
+    timing_info: Dict[str, TimeReport],
+    dt_atmos: float,
+):
+    sypd = get_sypd(timing_info, dt_atmos)
+    report = Report(setup=exp_info, times=timing_info, SYPD=sypd)
+    return dataclasses.asdict(report)
```

Comment:
if we already have the dataclass, why do we make it a dict here? why not let that be the responsibility of the function consuming the a dict?

Reply:
Good point, moving this to when the data is written, and remove `make_report`